### ISSUE TITLE
Make setup.py require dash

### ISSUE THREAD

**Issue by xhlulu** [plotly/dash-bio#135]
Really just a suggestion, but I believe it would be nice to have setup.py require dash by default:
https://github.com/plotly/dash-bio/blob/bd5ce53c4529ce3e0c783cb61eb6eab7b298dd93/setup.py#L11-L20

That way, when the user installs dash-bio, they simply need to create a venv and run:
```
pip install dash-bio
```
with dash, dash-html-components, dash_renderer being installed as well.

It could be implemented along the line of:

```python
 setup( 
     name=package_name, 
     version=package["version"], 
     author=package['author'], 
     packages=[package_name, '{}/utils'.format(package_name), '{}/component_factory'.format(package_name)], 
     include_package_data=True, 
     license=package['license'], 
     description=package['description'] if 'description' in package else package_name, 
     install_requires=[
         'dash', 'dash-html-components', 'dash_renderer'
     ] 
 ) 
```

**Comment by alexcjohnson** [comment]
Yes! See also https://github.com/plotly/dash-core/issues/24, seems to me once we do that with the core packages, auxiliary packages like dash-bio can simply require a known compatible dash version and everything else follows.

**Comment by shammamah** [comment]
@mkcor This seems like a pretty quick fix -- would you be able to work on it?

**Comment by mkcor** [comment]
Absolutely! Considering the time difference, I'll do it by tomorrow Tuesday.

**Comment by shammamah** [comment]
@mkcor That sounds like a great idea!

**Comment by mkcor** [comment]
[Requiring python~=3.3]
> @mkcor That sounds like a great idea!

Done! 156d436a0dd96794c0a1dc16b278af7ca49bbd21

**Comment by xhlulu** [comment]
Is dash tested for 3.3? I thought the tests were all done in 3.6 and 3.7.

**Comment by mkcor** [comment]
@xhlulu python~=3.3 means >=3.3 and <4 (as I wrote in the commit message).
Reference is still: https://packaging.python.org/guides/distributing-packages-using-setuptools/#python-requires

**Comment by alexcjohnson** [comment]
> python~=3.3 means >=3.3 and <4

@xhlulu is correct that we only explicitly test dash against 2.7, 3.6, and 3.7, so it's technically speculation to say that it works down to 3.3. However we do list it in dash's `setup.py`:

https://github.com/plotly/dash/blob/ab4f70849dfb70ece3ece42ec750075777fa8f22/setup.py#L43-L47

(hah, 3.7 isn't there, it should be!)

So I'd say 3.3 is safe, and if anyone reports a bug with a version <3.6 (in any of our dash repos) we can address it at that time.

**Comment by mkcor** [comment]
@alexcjohnson Sounds good.

We only test dash-bio against 3.6:

https://github.com/plotly/dash-bio/blob/ff372c4df10f350cbfb9f47a56aed1d9c569ade3/.circleci/config.yml#L84

Yes, we should also add 3.7:

https://github.com/plotly/dash-bio/issues/141

**Comment by xhlulu** [comment]
I'm reopening this issue to discuss which minimal version of dash, as well as other libraries, should be set for `setup.py`. For example, we currently have:
```python
...
    install_requires=[
        'biopython',
        'colour',
        'dash',
        'dash-html-components',
        'pandas',
        'parmed',
        'plotly',
        'scipy',
        'sklearn'
    ],
...
```

But there's no way to ensure that the version of `dash` or `sklearn` that the user has will work correctly with `dash-bio`. For example, if the user has a version of sklearn where k-means was not added, or where the methods used by `dash-bio` where not yet added, then `dash-bio` won't function correctly.

In this case should we set a minimal version for the important dependencies? e.g.
```python
...
    install_requires=[
        'biopython',
        'colour',
        'dash>=0.38.0',
        'dash-html-components',
        'pandas',
        'parmed',
        'plotly',
        'scipy',
        'sklearn>=0.20.1'
    ],
...
```
At the risk of creating installation conflicts for the user?

Tl;dr What minimal version of dash should `dash-bio` (and possibly `dash-cytoscape`) have in `install_requires`?

**Comment by mkcor** [comment]
Hi @xhlulu,

Thanks for bringing this up. It's probably okay to let dependencies be that 'abstract' in the `setup.py` file, although Dash Bio is more of an app than a library... For a discussion on this, please read: https://caremad.io/posts/2013/07/setup-vs-requirement/)

In any case, I would make sense to specify `sklearn>=0.20.1` at least in the `requirements.txt` file. Cc'ing @shammamah here.

**Comment by xhlulu** [comment]
Thanks for the link!

**Comment by shammamah** [comment]
Like I mentioned on #232, the Dash library has this similar setup with minimum versions specified in the `setup.py` file, but that seems to imply that it follows an "app" pattern rather than a "library" pattern. Should we change both libraries? cc @alexcjohnson

**Comment by mkcor** [comment]
It's not necessarily an issue (I mean, distributing a library that's actually an app). Using tools in ways they were not originally intended for is a major part of everything innovation. I think it's just interesting to have the discussion, learn about the background / context, and make informed decisions.

**Comment by alexcjohnson** [comment]
An interesting discussion indeed. The app/library and abstract/specific distinctions seem like good starting points for thinking through what will actually happen when a random user installs and uses this software, but they're not the end of the story.

An app, in this sense, means no further code is going to be connected to it, it's just going to be run as is. In that case locking down the exact versions of all the pieces makes sense: that's how the app was put together and tested, so stick with that.

`dash-bio` is fairly near the top of the stack, but it's still a library in the sense that it integrates with the user's code, which may independently import and use some of these same packages. We still test it with specific versions of all the dependents, so it would be nice if we could ensure that these same versions will always be used by everyone using it, but we can't, at least not in Python, where a given package can only have a single version in a given environment (as opposed to JS where the recursive `node_modules` structure allows each package to use its own versions of dependents... but that has its own problems!). So to play nice with our users' environments, we want to be flexible with versions of these dependencies.

That's fine for things like `pandas` which is widely used and (presumably) stable enough that whatever version a user has already will very likely work for us too. But for `dash` it's not fine: `dash` and `dash-bio` are being developed in tandem, we know there are breaking changes that preclude earlier versions, and it's completely reasonable for us to require users to upgrade `dash` when they upgrade `dash-bio`. So `pandas` is abstract, `dash` is specific, and every other dependent should be evaluated on a similar basis.

**Comment by xhlulu** [comment]
Thank you for the very clear breakdown!

**Comment by shammamah** [comment]
Closing this since it was fixed in https://github.com/plotly/dash-bio/pull/195.
